### PR TITLE
Create new Block on filter

### PR DESCRIPTION
**Pull request by dnhatn**
Today, we do not deduplicate data for filtered Block/Vector but rather share the underlying data with the existing one. This choice might improve performance since it avoids copying data. However, we have encountered issues due to the lack of clear ownership. We will likely introduce reference counting for Block/Vector. Until that is implemented, this PR addresses the duplication of data for filtered Block/Vector.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java [hunk 1]
```diff
@@ -45,7 +45,6 @@ protected AbstractBlock(int positionCount, @Nullable int[] firstValueIndexes, @N
         this.blockFactory = blockFactory;
         this.firstValueIndexes = firstValueIndexes;
         this.nullsMask = nullsMask == null || nullsMask.isEmpty() ? null : nullsMask;
-        assert (firstValueIndexes == null && this.nullsMask == null) == false;
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java:48]: This assertion prevents me from creating a valid Block when both nullsMask and firstValueIndexes are null. I've removed this assertion. I think the correct one is `nullMask != null -> firstValueIndexes == null || firstValueIndexes.length == positionCount`. I will need to modify some tests before enabling that assertion, and I will do this in a follow-up.

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/DoubleBigArrayVector.java [hunk 2]
```diff
@@ -60,7 +60,11 @@ public long ramBytesUsed() {
 
     @Override
     public DoubleVector filter(int... positions) {
-        return new FilterDoubleVector(this, positions);
+        final DoubleArray filtered = blockFactory.bigArrays().newDoubleArray(positions.length, true);
+        for (int i = 0; i < positions.length; i++) {
+            filtered.set(i, values.get(positions[i]));
+        }
+        return new DoubleBigArrayVector(filtered, positions.length, blockFactory);
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/DoubleBigArrayVector.java:67]: I think we don't account the memory usage of BigArrayVector correctly. However, we aren't using them now. I can fix them in a follow-up.

Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/DoubleBigArrayVector.java:67]: +++

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java (hunk 1 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java:48]: If there aren't any nulls or multivalued fields shouldn't we make a `Vector` instead?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/nulls/Coalesce.java [hunk 3]
```diff
@@ -149,15 +150,17 @@ public Block.Ref eval(Page page) {
                         1,
                         IntStream.range(0, page.getBlockCount()).mapToObj(b -> page.getBlock(b).filter(positions)).toArray(Block[]::new)
                     );
-                    for (EvalOperator.ExpressionEvaluator eval : evaluators) {
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/nulls/Coalesce.java:152]: :+1:

This strategy is *even* worse with this work around of copying to filter. Oh well, we know coalesce and case ain't fast yet.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java (hunk 1 above)
Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java:48]: Thanks, I restored the assertion with the message in https://github.com/elastic/elasticsearch/pull/100465/commits/97760a7ff6a791552e2369af16853f700ba817a1.

**Comment by dnhatn** [comment]
run lasticsearch-ci/part-2

**Comment by dnhatn** [comment]
run elasticsearch-ci/part-2

**Comment by dnhatn** [comment]
The PR is ready for review, despite two test failures, as they are likely test issues.

**Review by costin — APPROVED** [review]
LGTM.

**Comment by dnhatn** [comment]
```
REPRODUCE WITH: ./gradlew ':x-pack:plugin:esql:qa:server:multi-node:javaRestTest' --tests "org.elasticsearch.xpack.esql.qa.multi_node.EsqlSpecIT" -Dtests.method="test {rename.DuplicateProjectEval}" -Dtests.seed=6FDA6166F8A1A75C -Dtests.locale=nl -Dtests.timezone=Europe/Zagreb -Druntime.java=21

org.elasticsearch.xpack.esql.qa.multi_node.EsqlSpecIT > test {rename.DuplicateProjectEval} FAILED
    org.junit.ComparisonFailure: expected:<[2]> but was:<[4]>
        at __randomizedtesting.SeedInfo.seed([6FDA6166F8A1A75C:E78E5EBC565DCAA4]:0)
```

This seems unrelated to this PR. I will look into this later.

@elasticmachine run elasticsearch-ci/part-2

**Comment by costin** [comment]
A nice side effect of this PR is removing close to 1500 LOC.

**Review by luigidellaquila — COMMENTED** [review]
Thanks Nhat, this will make block management much easier.
There is a single case where this could have a significant negative impact though, that is `mv_expand`, where we used FilterBlocks to avoid duplicating data. With this change, mv_expand will very likely create huge blocks/pages; IMHO we should at least track this problem with an issue and fix it asap, maybe returning multiple pages and building them lazily in case the positions array is much bigger than the original block size (or just above a defined threshold), something like

```
public Iterator<XBlock> filter(int... positions) {
```

[edit] well, probably we should leave `Block.filter()` API as it is and just avoid to use it in MvExpandOperator

**Review by luigidellaquila — APPROVED** [review]

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/DoubleBigArrayVector.java (hunk 2 above)
Comment (ChrisHegarty) [reply x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/DoubleBigArrayVector.java:67]: ++

**Review by ChrisHegarty — APPROVED** [review]
LGTM

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/FilterLongVector.java [hunk 4]
```diff
@@ -1,105 +0,0 @@
...
-
-    private final LongVector vector;
-
-    private final LongBlock block;
-
-    FilterLongVector(LongVector vector, int... positions) {
-        super(positions, vector.blockFactory());
-        this.vector = vector;
-        this.block = new LongVectorBlock(this);
-    }
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/FilterLongVector.java:29]: ++ to removing the `X-FilterVector`s; the way their references behaved made ownership hard to reason about.

thought: Once we introduce refcounting, I think we'll want to bring back the essential logic from this, i.e. take an existing vector + a positions array/bitmap to create a filtered, shallow copy of an existing vector. But that can probably go directly into `X-Vector` or `X-ArrayVector` then.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/FilterOperator.java [hunk 5]
```diff
@@ -71,10 +71,18 @@ protected Page process(Page page) {
...
+            try {
+                for (int i = 0; i < page.getBlockCount(); i++) {
+                    filteredBlocks[i] = page.getBlock(i).filter(positions);
+                }
+                success = true;
+            } finally {
+                Releasables.closeExpectNoException(page::releaseBlocks);
+                if (success == false) {
+                    Releasables.closeExpectNoException(filteredBlocks);
+                }
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/FilterOperator.java:84]: Nice, this is much easier to reason about!

This should make the fix suggested in https://github.com/elastic/elasticsearch/pull/100278 obsolete. To confirm, you could remove `FilterOperatorTests::canLeak` and see if `FilterOperatorTests::testSimpleSmallInput` and ...::testSimpleLargeInput` still pass - or I can make a small follow-up PR that does that.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/LimitOperator.java [hunk 6]
```diff
@@ -92,8 +93,18 @@ public Page getOutput() {
...
+            try {
+                for (int b = 0; b < blocks.length; b++) {
+                    blocks[b] = lastInput.getBlock(b).filter(filter);
+                }
+                success = true;
+            } finally {
+                Releasables.closeExpectNoException(lastInput::releaseBlocks);
+                lastInput = null;
+                if (success == false) {
+                    Releasables.closeExpectNoException(blocks);
```
Comment (bpintea) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/LimitOperator.java:106]: Could this (and the similar in `FilterOperator`) not triger a NPE if  `blocks` doesn't get to be fully initialized?

**Review by bpintea — APPROVED** [review]
Nice!
I guess https://github.com/elastic/elasticsearch/issues/99997 can be linked and closed.

**Review by nik9000 — APPROVED** [review]

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/FilterOperator.java (hunk 5 above)
Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/FilterOperator.java:84]: Good point, I've disabled `allowLeaking` in both Limit and Filter operators in https://github.com/elastic/elasticsearch/pull/100465/commits/e1e7036cecb8d00e5b05ee66b346f8ce122b5bda

**Comment by dnhatn** [comment]
Thanks everyone for the review. I will wait for CI to be green and merge.

**Pull request merged**